front50-azure/src/main/java/com/netflix/spinnaker/front50/model/AzureStorageService.java

```diff
@@ -0,0 +1,319 @@
+      writeLastModified(objectType.group);
+      log.info("{} object {} for  has been successfully uploaded.", objectType.group, key);
+    }
+    catch (StorageException se) {
+      logStorageException(se, key);
+    }
+    catch (Exception e)
+    {
+      log.error("Error encountered attempting to store {}: {}", key, e.getMessage());
+    }
+  }
+
+  @Override
+  public Map<String, Long> listObjectKeys(ObjectType objectType) {
+    String path = rootFolder + "/" + objectType.group;
+    Map<String, Long> objectKeys = new HashMap<>();
+    try {
+      ResultContinuation token = null;
+      do {
+        ResultSegment<ListBlobItem> result = getBlobContainer().listBlobsSegmented(path, true, null, 10000, token, null, null);
+        token = result.getContinuationToken();
+
+        result.getResults().stream().filter(item -> match(item, objectType.defaultMetadataFilename)).forEach(item -> {
+          CloudBlob blob = (CloudBlob) item;
+          DateTime modDate = new DateTime(blob.getProperties().getLastModified());
+          objectKeys.put(getBlobKey(objectType, blob.getUri().toString()), modDate.getMillis());
+        });
+
+      } while (token != null);
+    } catch (StorageException se) {
+      logStorageException(se, "");
+    } catch (Exception e) {
+      log.error("Failed to retrieve objects from {}: {}", path, e.getMessage());
+    }
+    return objectKeys;
+  }
+
+  @Override
+  public <T extends Timestamped> Collection<T> listObjectVersions(ObjectType objectType,
+                                                                  String objectKey,
+                                                                  int maxResults) throws NotFoundException {
+    Set<T> results = new HashSet<>();
+    String fullKey = buildKeyPath(objectType.group, objectKey, objectType.defaultMetadataFilename);
+    try {
+      ResultContinuation token = null;
+      EnumSet<BlobListingDetails> listDetails = EnumSet.of(BlobListingDetails.SNAPSHOTS);
+      do {
+        ResultSegment<ListBlobItem> result = getBlobContainer().listBlobsSegmented(fullKey, true, listDetails, maxResults, token, null, null);
+        token = result.getContinuationToken();
+        for (ListBlobItem item : result.getResults()) {
+          CloudBlockBlob blob = (CloudBlockBlob)item;
+          T blobObject = deserialize(blob, (Class<T>) objectType.clazz);
+          blobObject.setLastModified(blob.getProperties().getLastModified().getTime());
+          results.add(blobObject);
+        }
+
+      } while (token != null);
+    } catch (StorageException se) {
+      logStorageException(se,fullKey);
+    } catch (Exception e) {
+      log.error("Error retrieving versions for {} object: {}", fullKey, e.getMessage());
+    }
+    return results;
+  }
+
+  @Override
+  public long getLastModified(ObjectType objectType) {
+    try {
+      CloudBlockBlob lastMod = getBlobContainer().getBlockBlobReference(getLastModifiedFile(objectType.group));
+      if (lastMod.exists()) {
```

Comment:
FYI, for us there was a 404 that showed up in the users API stats when the file isnt there. Since this is polling so much and so often, there were lots of 404 noise. That's why I created the timestamp file if it wasnt already there. I dont know how the underlying Azure call behaves or reports. The code here looks like it is good, but you might want to check.